### PR TITLE
Add OME-NGFF Reader

### PR DESCRIPTION
**Pull request by DavidStirling**
This PR adds a new reader for the [OME-NGFF](https://ngff.openmicroscopy.org/latest/) file format. This should be able to handle both standard and HCS varients of the format. It _may_ also be able to read other .zarr image files, depending on their contents. Compatibility is evaluated based on the presence of '.zarr' in a directory name.

Determining the contents of an NGFF file can be tricky. This implementation works with a variety of zarrs in both older versions and newer versions of the spec. I anticipate that further updates will be needed once V4 of the specification releases, but those changes should help to improve efficiency when working with S3. For cloud work, it's currently best to run the `zarr.convenience.consolidate_metadata` function on zarr images before uploading them to S3, which makes it possible to read their contents with substantially fewer S3 requests. Spec V4 should provide some better solutions to this.

I expect the reader to need some tweaks to deal with edge cases and address bugs.

I'll be making another PR with some revisions to the build system, but for now packing this seems to work smoothly with the addition of the following dependencies: `zarr`, `fsspec` and `lxml`. These are fairly lightweight. The latter can probably be dropped once the format matures and OME-XML is fully redundant.

This reader operates in pure Python, so there is no need for java.

I've also moved the `find_scale_to_match_bioformats` method from the imageio reader into the core reader class. It's going to be useful in a few scenarios (including this).

### CODE REVIEW

**Comment by DavidStirling** [comment]
For testing purposes, existing image sets can be converted to .zarr with [NGFF-Converter](https://downloads.glencoesoftware.com/public/NGFF-Converter/latest/)

File: cellprofiler_core/readers/ngff_reader.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+FORMAT_TESTER = re.compile(r"\.zarr([\\/]|$)", flags=re.IGNORECASE)
+
+LOGGER = logging.getLogger(__name__)
+
+
+class NGFFReader(Reader):
+    """ Derive from this abstract Reader class to create your own image reader in Python
+
+    You need to implement the methods below in the derived class.
+    """
```
Comment (gnodar01) [inline cellprofiler_core/readers/ngff_reader.py:25]: Should change to something like (I need to do this on the Bio-Formats reader as well):

```
    """
    Reader for zarr files, specifically those following the OME-NGFF file format,
    although has some support for other zarr image files, depending on their contents.
    """
```

File: setup.py [hunk 2]
```diff
@@ -34,13 +36,14 @@
...
+        "zarr>=2.10",
     ],
     license="BSD",
     name="cellprofiler-core",
     package_data={"cellprofiler_core": ["py.typed"]},
     packages=setuptools.find_packages(exclude=["tests"]),
     python_requires=">=3.8",
     url="https://github.com/CellProfiler/core",
-    version="4.2.4",
+    version="5.0.0",
```
Comment (gnodar01) [inline setup.py:47]: `CellProfiler/setup.py` should be changed to `cellprofiler-core==5.0.0` in CellProfiler/CellProfiler#4660

**Review by gnodar01 — CHANGES_REQUESTED** [review]

File: setup.py (hunk 2 above)
Comment (DavidStirling) [reply setup.py:47]: Done, thanks

File: cellprofiler_core/readers/ngff_reader.py (hunk 1 above)
Comment (DavidStirling) [reply cellprofiler_core/readers/ngff_reader.py]: Well spotted, thanks!

**Review by gnodar01 — APPROVED** [review]

**Pull request merged**